posts/2022/08-25-adapter_pattern.md

```diff
@@ -0,0 +1,252 @@
+
+
+# 2. The puzzle pieces
+
+
+Before I start writing any code, I like to think about which pieces I will need to put in place to complete the task.
+In the case of our hockey games, we will need a supervised process that ensures that the APIs are queried periodically, code for making the API calls, code for standardizing the return data and last but not least the database components, namely a table and a schema for the match data.
+
+
+# 3. Arranging the pieces
+
+
+Now that we know what pieces we need, let's figure out how to put them together. The database side is not different from any other project, so we'll just consider it done. 
+
+Next up is our process. We could use a task that triggers the API calls, stores the fetched data in the database and then calls itself again after we run `Process.sleep/1` with the required amount n of milliseconds. However, since the API calls and data processing operations also take a small amount of time, we would end up querying the API not every n milliseconds but instead every n milliseconds + the time required to do all the other stuff.
+A GenServer is more reliable since we can utilize `Process.send_after/4` to have the GenServer call itself reliably every n milliseconds.
+
+Since the GenServer is doing the same things for both APIs, we can start the same GenServer twice; all we have to do is register the two instances of our GenServer under different names.
+
+We also need code for calling our APIs. I decided to create one API module per queried API. While the two API modules have the same job (calling the API, fetching JSON data, decoding JSON data, standardizing data, returning standardized to the GenServer), they are operating under different circumstances: different HTTP addresses, plus the fact that API B also takes in the additional search parameter. 
+
+This is the big moment: enter Behaviour.
+
+Since we assume that both our API modules will perform the same actions and send data in the same unified format back to the GenServer, we can go one step further and not just assume but require that they do so. 
+To this end, we can use another module to define the desired Behaviour and implement said Behaviour in our API modules. Now, we cannot only be sure that our existing API modules act in a certain way, but we can also easily add further API modules in the future. All we have to do is implement the same Behaviour for them, thus ensuring that we don't accidentally add a rogue API module that acts different from the rest and risks bringing down its associated GenServer.
+
+But enough with the talk. Let's look at some code!
+
+
+# 4. The code
+
+
+**General organization**
+
+```
+
+lib
+│   ├── my_app
+│   │   ├── application.ex
+│   │   ├── config.ex
+│   │   ├── match_data_processors
```

Comment:
Are they actually data processors? Or just clients?

Reply:
I leave them as processors. they query and convert after all